torchopt/_src/linalg.py

```diff
@@ -0,0 +1,219 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+# This file is modified from:
+# https://github.com/google/jax/blob/main/jax/_src/scipy/sparse/linalg.py
+# ==============================================================================
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+# pylint: disable=invalid-name
+
+from functools import partial
+from typing import Any, Callable, List, Optional, Sequence, Union
+
+import torch
+
+from torchopt._src.typing import TensorTree
+from torchopt._src.utils import pytree
+
+
+# Aliases for working with pytrees
+def _vdot_real_part(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+    x = x.view(-1)
+    y = y.view(-1)
+    result = torch.dot(x.real, y.real)
+    if x.is_complex() or y.is_complex():
+        result += torch.dot(x.imag, y.imag)
+    return result
+
+
+def _vdot_real_tree(x: TensorTree, y: TensorTree) -> TensorTree:
+    return pytree.tree_map(_vdot_real_part, x, y)
+
+
+def _identity(x: TensorTree) -> TensorTree:
+    return x
+
+
+def _normalize_matvec(
+    f: Union[Callable[[TensorTree], TensorTree], torch.Tensor]
+) -> Callable[[TensorTree], TensorTree]:
+    """Normalize an argument for computing matrix-vector products."""
+    if callable(f):
+        return f
+
+    assert isinstance(f, torch.Tensor)
+    if f.ndim != 2 or f.shape[0] != f.shape[1]:
+        raise ValueError(f'linear operator must be a square matrix, but has shape: {f.shape}')
+    return partial(torch.matmul, f)
+
+
+def _safe_sum(obj: Union[Any, Sequence[Any]]) -> Any:
+    if isinstance(obj, (list, tuple)):
+        return sum(obj)
+    return obj
+
+
+# pylint: disable-next=too-many-locals
+def _cg_solve(
+    A: Callable[[TensorTree], TensorTree],
+    b: TensorTree,
+    x0: TensorTree,
+    *,
+    maxiter: int,
+    rtol: float = 1e-5,
+    atol: float = 0.0,
+    M: Callable[[TensorTree], TensorTree] = _identity,
+) -> TensorTree:
+    # https://en.wikipedia.org/wiki/Conjugate_gradient_method#The_preconditioned_conjugate_gradient_method
+
+    # tolerance handling uses the "non-legacy" behavior of `scipy.sparse.linalg.cg`
+    bs = _safe_sum(_vdot_real_tree(b, b))
+    atol2 = max(rtol**2 * bs, atol**2)
```

Comment:
`bs` could be a tree of tensors (e.g. `dict` of tensors). Comparing to float can be failed.